src/mgba_printf/mgba.c

```diff
@@ -0,0 +1,181 @@
+        else if (*src >= CHAR_A && *src <= CHAR_Z)
+            modifiedCode = *src-(CHAR_A-'A'); // upper-case characters
+        else if (*src >= CHAR_0 && *src <= CHAR_9)
+            modifiedCode = *src-(CHAR_0-'0'); // numbers
+        else if (*src == CHAR_SPACE)
+            modifiedCode = ' '; // space
+        else if (*src == CHAR_EXCL_MARK)
+            modifiedCode = '!'; // exclamation point
+        else if (*src == CHAR_QUESTION_MARK)
+            modifiedCode = '?'; // question mark
+        else if (*src == CHAR_PERIOD)
+            modifiedCode = '.'; // period
+        else if (*src == CHAR_DBL_QUOTE_LEFT || *src == CHAR_DBL_QUOTE_RIGHT)
+            modifiedCode = '"'; // double quote
+        else if (*src == CHAR_SGL_QUOTE_LEFT || *src == CHAR_SGL_QUOTE_RIGHT)
+            modifiedCode = '\''; // single quote
+        else if (*src == CHAR_CURRENCY)
+            modifiedCode = '$'; // currency mark (pokemonies in game, dollar sign in logs)
+        else if (*src == CHAR_COMMA)
+            modifiedCode = ','; // comma
+        else if (*src == CHAR_MULT_SIGN)
+            modifiedCode = '#'; // pound, hashtag, octothorpe, whatever
+        else if (*src == CHAR_SLASH)
+            modifiedCode = '/'; // slash
+        else if (*src == CHAR_LESS_THAN)
+            modifiedCode = '<'; // less than sign
+        else if (*src == CHAR_GREATER_THAN)
+            modifiedCode = '>'; // greater than sign
+        else if (*src == CHAR_PERCENT)
+            modifiedCode = '%'; // percentage
+        else if (*src == CHAR_LEFT_PAREN)
+            modifiedCode = '('; // opening parentheses
+        else if (*src == CHAR_RIGHT_PAREN)
+            modifiedCode = ')'; // closing parentheses
+        dst[i] = modifiedCode;
+    }
+    dst[i] = 0;
+    return i;
+}
+
+static void MgbaPrintfBounded(s32 level, const char* ptr, ...)
+{
+    va_list args;
+
+    level &= 0x7;
+    va_start(args, ptr);
+    mini_vsnprintf(REG_DEBUG_STRING, MGBA_REG_DEBUG_MAX, ptr, args);
+    va_end(args);
+    *REG_DEBUG_FLAGS = level | 0x100;
+}
+
+void MgbaPrintf(s32 level, const char* ptr, ...)
+{
+    va_list args;
+    u32 offset = 0;
+    u32 n = 0;
+    u32 i;
+    char *buffer = Alloc(PRINTF_BUFFER_SIZE);
+
+    MGBA_ASSERT(buffer != NULL);
+
+    level &= 0x7;
+
+    va_start(args, ptr);
+    n = mini_vsnprintf(buffer, PRINTF_BUFFER_SIZE, ptr, args);
+    va_end(args);
+
+    MGBA_ASSERT(n < PRINTF_BUFFER_SIZE);
+
+    do 
+    {
+        for (i = 0; i < MGBA_REG_DEBUG_MAX; ++i)
+        {
+            REG_DEBUG_STRING[i] = buffer[offset + i];
+            if (buffer[offset + i] == 0)
+                break;
+        }
+        offset += i;
+        *REG_DEBUG_FLAGS = level | 0x100;
+    } while ((i == MGBA_REG_DEBUG_MAX) && (buffer[offset] != '\0'));
+
+    Free(buffer);
+}
+
+bool8 MgbaOpen(void)
+{
+    *REG_DEBUG_ENABLE = 0xC0DE;
+    return *REG_DEBUG_ENABLE == 0x1DEA;
+}
+
+void MgbaClose(void)
```

Comment:
this should be optional and stubbed if mgba debugging is disabled, the HW registers do not work on a real gba or anything that is not mgba.